cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -1343,27 +1343,71 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
+                        Opcode::Fneg => (0x8000000000000000, SseOpcode::Xorpd),
+                        _ => unreachable!(),
+                    },
+                    _ => panic!("unexpected type {:?} for Fabs", output_ty),
+                };
 
-            for inst in Inst::gen_constant(dst, val, output_ty, |reg_class, ty| {
-                ctx.alloc_tmp(reg_class, ty)
-            }) {
-                ctx.emit(inst);
-            }
+                for inst in Inst::gen_constant(dst, val, output_ty, |reg_class, ty| {
+                    ctx.alloc_tmp(reg_class, ty)
+                }) {
+                    ctx.emit(inst);
+                }
+
+                ctx.emit(Inst::xmm_rm_r(opcode, src, dst));
+            } else {
+                // Eventually vector constants should be available in `gen_constant` and this block
+                // can be merged with the one above (TODO).
+                if output_ty.bits() == 128 {
+                    // Move the `lhs` to the same register as `dst`; this may not emit an actual move
+                    // but ensures that the registers are the same to match x86's read-write operand
+                    // encoding.
+                    match src {
+                        RegMem::Reg { reg } => ctx.emit(Inst::gen_move(dst, reg, output_ty)),
+                        RegMem::Mem { .. } => unimplemented!(),
+                    };
+
+                    // Generate an all 1s constant in an XMM register.
+                    let tmp = ctx.alloc_tmp(RegClass::V128, output_ty);
+                    let cond: FloatComparisonPredicate = FloatCC::Equal.into();
+                    let cmpps = Inst::xmm_rm_r_imm(
+                        SseOpcode::Cmpps,
+                        RegMem::reg(tmp.to_reg()),
+                        tmp,
+                        cond.encode(),
+                    );
+                    ctx.emit(cmpps);
+
+                    // Shift the all 1s constant to generate the mask.
+                    let lane_bits = output_ty.lane_bits();
+                    let shift_by = RegMemImm::imm((lane_bits - 1) as u32);
+                    let (shift_opcode, opcode) = match (op, lane_bits) {
+                        (Opcode::Fabs, 32) => (SseOpcode::Psrld, SseOpcode::Andps),
+                        (Opcode::Fabs, 64) => (SseOpcode::Psrlq, SseOpcode::Andpd),
+                        (Opcode::Fneg, 32) => (SseOpcode::Psllq, SseOpcode::Xorps),
+                        (Opcode::Fneg, 64) => (SseOpcode::Psllq, SseOpcode::Xorpd),
+                        _ => unreachable!(),
+                    };
+                    let shift = Inst::xmm_rmi_reg(shift_opcode, shift_by, tmp);
+                    ctx.emit(shift);
 
-            ctx.emit(Inst::xmm_rm_r(opcode, src, dst));
+                    // Invert the sign bit.
+                    let xor = Inst::xmm_rm_r(opcode, RegMem::reg(tmp.to_reg()), dst);
```

Comment:
nit: the above comment and this variable name hide the fact that it can be used for AND'ing and taking the absolute value too.